### PR TITLE
Make sof_compr_copy work for capture direction

### PR DESCRIPTION
**Pull request by LaurentiuM1234**
Because of the fact that `sof_compr_copy()` was only copying data from user space this means that only the playback direction would work on compress streams. What these patches do is enable `sof_compr_copy()` to copy data from/to user space depending on the direction: playback or capture.

### CODE REVIEW

File: sound/soc/sof/compress.c [hunk 1]
```diff
@@ -296,30 +296,40 @@ static int sof_compr_trigger(struct snd_soc_component *component,
...
 	n = rtd->buffer_size - offset;
 
 	if (count < n) {
-		ret = copy_from_user(ptr, buf, count);
+		if (copy_from_user(ptr, buf, count))
+			return -EFAULT;
 	} else {
-		ret = copy_from_user(ptr, buf, n);
-		ret += copy_from_user(rtd->dma_area, buf + n, count - n);
+		if (copy_from_user(ptr, buf, count))
```
Comment (plbossart) [inline sound/soc/sof/compress.c]: you are replacing 

copy_from_user(ptr, buf, n);

by 

copy_from_user(ptr, buf, count))

is this intentional?

File: sound/soc/sof/compress.c [hunk 2]
```diff
@@ -296,30 +296,40 @@ static int sof_compr_trigger(struct snd_soc_component *component,
...
+			return -EFAULT;
+
+		if (copy_from_user(rtd->dma_area, buf + n, count - n))
```
Comment (plbossart) [inline sound/soc/sof/compress.c]: I don't get why we moved from accumulating the bytes received to testing from something different from zero. Even if this was desired, it would need to be done in a separate patch prior to the introduction of a new function

Comment (paulstelian97) [reply sound/soc/sof/compress.c]: I have to agree with Pierre here, I can also see some changes that are unsound... I need a proper explanation before ever accepting them.

File: sound/soc/sof/compress.c (hunk 1 above)
Comment (LaurentiuM1234) [reply sound/soc/sof/compress.c]: It's not. Will have to fix it. Thanks for pointing it out!

File: sound/soc/sof/compress.c (hunk 2 above)
Comment (LaurentiuM1234) [reply sound/soc/sof/compress.c]: The reason I changed this is because I believe `sof_compr_copy` should return something negative in case of error so compress_offload API can know that the copy operation went wrong. From my understanding, in the previous version, `sof_compr_copy` would return something >= 0 even if the copy_{from/to}_user failed. Although this would mean that a partial copy was done, wouldn't it be safer to have a copy of count size or nothing?

**Comment by LaurentiuM1234** [comment]
Update:

1) We no longer test the return value of copy_{from/to}_user. The initial version was working just fine so there was really no need for this change. This way we can also allow partial copy.

**Review by paulstelian97 — APPROVED** [review]
Now it looks more like it.

**Comment by dbaluta** [comment]
@LaurentiuM1234 I think 'move' patch should just do that. Move code from one place to another, in preparation for next patch. 

No modifications. We do that in a follow up patch. So that the code is always clear and the intention of the patch is obvius :)

**Comment by paulstelian97** [comment]
> I think 'move' patch should just do that. Move code from one place to another, in preparation for next patch.

I think it's doing that already at this moment.

**Comment by LaurentiuM1234** [comment]
> @LaurentiuM1234 I think 'move' patch should just do that. Move code from one place to another, in preparation for next patch.
> 
> No modifications. We do that in a follow up patch. So that the code is always clear and the intention of the patch is obvius :)

The patch does just that but yeah I agree it doesn't look that way. Maybe because `sof_compr_copy_playback` starts at the line `sof_compr_copy` used to so git thinks I've replaced `sof_compr_copy` with it.

**Review by dbaluta — APPROVED** [review]
This looks good to me now. Thanks @LaurentiuM1234 !

File: sound/soc/sof/compress.c (hunk 2 above)
Comment (dbaluta) [reply sound/soc/sof/compress.c]: @plbossart marking this as resolved after @LaurentiuM1234 last comment. We revert back to only move code around and add capture.

**Comment by dbaluta** [comment]
would love to have an approval from Intel on this.
